### PR TITLE
Ctrl plus scroll to change playback rate

### LINKED ISSUE #616: [Feature Request] Variable playback speed with smaller step
**Is your feature request related to a usage problem (not a bug)? Please describe.**
Variable playback speed is great for speeding up videos, however a 0.25 step is quite big when you want to find the perfect speed to match exactly the speach in the video (especially long ones such as conferences and talks)

**Describe the solution you'd like to see implemented**
There could be an option in the settings to set the step, so that if you select 0.1, you'll be able to choose 0.8, 0.9, 1, 1.1, 1.2, etc. However, this could lead to too many choices for small steps.

Another solution could be using buttons such as the ones on the toolbar of the browser extension _Enhancer for Youtube_ (see screenshot below). Clicking on minus/plus will decrease/increase the playback speed (by a step defined in settings), and clicking on the middle button will set back the playback speed to 1.
If only a few people are interested in such a feature, a contextual menu option would be enough.

**Screenshots**
Capture d’écran 2020-10-08 à 10 54 35

**Additional context**
Chrome addon : https://chrome.google.com/webstore/detail/enhancer-for-youtube/ponfpcnoihfmfllpaingbgckeeldkhle
Firefox addon : https://addons.mozilla.org/en-US/firefox/addon/enhancer-for-youtube/

### LINKED ISSUE #1238: Custom playback speed option
In current moment playback speed possible to be or 1.25 or 1.5. 
Please add more flexibility in playback speeds settings. It will be nice to have possibility to change playback speed with 0.1 steps. 
Also will be nice to change playback speed pointing to playback speed control and using mouse wheel.

### LINKED ISSUE #1713: Feature Request - Use ctrl+scroll for playback speed
Allow for playback speed control using ctrl+scrollwheel while your mouse cursor is over the video. A ctrl+click on the video will quickly return it back to 1x.

This is a feature that is in the popular "Enhancer for YouTube" browser plugin for Chrome and Firebox. This is super helpful for long tutorial-based videos as it allows you to ramp the speed up and down quickly while scrubbing through the video. 

image

### PR DESCRIPTION
**Pull request by codysechelski**
---
Ctrl plus scroll to change playback rate
---

**Important note**
We may remove your pull request if you do not use this provided PR template correctly.

**Pull Request Type**
Please select what type of pull request this is:
- [ ] Bugfix
- [x] Feature Implementation

**Related issue**
Please link the issue your pull request is referring to. If this pull request fully resolves the relevant issue, put "closes" before the issue number. Example: "closes #123456".
closes #1713 
closes #616
closes #1238

**Description**
Please write a clear and concise description of what the pull request does.
- [x] Adds the ability to use the ctrl key and mouse wheel to quickly change the playback rate.
- [x] Changes the max playback rate from `3` to `8`
- [x] Adds the ability to use ctrl+click to return to the default playback rate.
- [x] Adds the ability to enable/disable the above in the `settings > player settings` screen (defaults to disabled)

**Screenshots (if appropriate)**
Please add before and after screenshots if there is a visible change.
Added Setting
image
Added Tooltip for Setting
image
Showing clicking without the control key pressed (the behavior is unchanged)
9E06A815-9340-48D2-A1AD-6406B1463940
Showing Ctrl+Scrolling Up, Ctrl+Click, Ctrl+ Scrolling Down (NOTE: the key display software that I'm using does not support showing wheel events. You're just going to have to trust me.)
A91EBEA4-5854-494B-87DD-7AB48BA255D2

**Testing (for code that is not small enough to be easily understandable)**
Has this pull request been tested?
Please describe shortly how you tested it and whether there are any ramifications remaining. 

Manually tested:
* enabling/disabling the feature
  - No apparent user impact when disabled
  - Enabling the feature works as expected
* Scrolling up and down without holding ctrl
  - When "scroll to change volume" is turned on, it functions as expected
  - When "scroll to change volume" is turned off, it functions as expected
* Scrolling up and down while holding ctrl
  - When "scroll to change volume" is turned on, it functions as expected
  - When "scroll to change volume" is turned off, it functions as expected
* Ctrl+Click
  - This works as expected but there may be a better way to do this. See my first comment on this PR for more details

**Desktop (please complete the following information):**
  - OS: [e.g. iOS] : Windows 10 Home
  - OS Version: [e.g. 22]: Version 21H1, Build 19043.1237 with Windows Feature Experience Pack 120.2212.3530.0
  - FreeTube version: [e.g. 0.8] 0.14.0

**Additional context**

### CODE REVIEW

**Comment by codysechelski** [comment]
Ok, so here are some important points I want to make about this PR...

1. Having the ability to "ramp up" and "ramp down" playback speed will often illicit a different behavior than just having a button to change the playback speeds.

In my opinion, having to locate the button, click it, and select the new playback rate is a bit tedious. Not bad, just tedious. So I think a lot of users would use this as a "set it and forget it" option for a given video. For example, I want to watch a 30-minute video about how coffee is made, but I don't have time. I would likely set the video to play at 1.25x or 1.5x speed and leave it like that for the entire video.

However, if I can effectively turn my scroll wheel into a "jog wheel" to scrub through videos, that changed the way I will use it. For example, If I want to "fast forward" through the sponsored portion of a video, or "scrub" through a video that I have seen before to find a particular spot. Personally, I use it a lot when watching tutorial videos. There are often parts of what they are showing that I already know how to do and I just want to "get to the good part"

I say all of this to justify why I believe that the max playback rate should be changed from 3x to 8x. With a max of 3x, you don't really get a lot of the benefits that I stated above. Perhaps we can keep the max 3 when the "scroll to change playback rate" is disabled and make it 8 when you turn it on, but I don't see any other way to control the playback rate other than the button in the control bat which limits you to a district list of speeds. So is it work adding that logic?

2. As I just stated, the playback rate button has a district list of speeds to choose from. I don't think it makes sense to show every speed that is achievable using ctrl+scoll in that list. It would not be a good user experience. However, ctrl+scrolling does change the label of that button to the current playback rate. At first, I thought it might be confusing for a user to see a speed on the button label that was not available in the popup list. But after using it, it does not seem confusing to me. I would love someone else's opinion on this.

I thought perhaps we could disable showing the popup menu when this feature was turned on, but that kind of seems like take a feature away without much gain. thoughts?

3. I have 2 points to make about the ctrl+click to return to "normal" playback speed. The first point is this, I originally had it to where a ctrl+click would return you back to 1x. Then I thought about it some more and decided that it would be better to return them to the default playback rate that they have selected in settings, which I think for most users will be the default 1x. What are yall's thoughts on this?

4. Lastly, let's talk about how I am achieving the ctrl+click functionality. The native behavior when clicking on the video is toggling play/pause. This happens no matter if the ctrl key is pressed. I thought I would be able to override this but it seems that it is part of the video,js library and after looking at the docs, I don't see a way to override this. Perhaps a better developer can jump in and validate that.

So if you look at the fl-video-player.js, around line 530, you will see what I did... The click will pause the video, so I check to see it's it paused and if the ctrl key was pressed, I just play it again. It worked better than I expected it, It's actually a really smooth transition. The only thing is that the play button will sometimes flash for a millisecond or so, o I just removed it and restoe4d it back afterwards. This seems a bit hacky to me, but what do yall think?

Ok, so sorry to post such a wall of text and thank you to whoever read all of this. 😁

**Comment by jasonhenriquez** [comment]
This was quick! I will not be able to look at this in detail for the time being, but I will try to address what you've said now.

> I say all of this to justify why I believe that the max playback rate should be changed from 3x to 8x which I did in the PR. With a max of 3x, you don't really get a lot of the benefits that I stated above. Perhaps we can keep the max 3 when the "scroll to change playback rate" is disabled and make it 8 when you turn it on, but I don't see any other way to control the playback rate other than the button in the control bar which limits you to a district list of speeds. So is it work adding that logic?

There is a bit more work required to do this right. You will have to update the playbackRates array in ft-video-player.js with the new values. You will also have to change the max value to 8 in player-settings.vue for default playback rate. Also, you should have this feature change in your main summary.

>     1. As I just stated, the playback rate button has a district list of speeds to choose from. I don't think it makes sense to show every speed that is achievable using ctrl+scoll in that list. It would not be a good user experience. However, ctrl+scrolling does change the label of that button to the current playback rate. At first, I thought it might be confusing for a user to see a speed on the button label that was not available in the popup list. But after using it, it does not seem confusing to me. I would love someone else's opinion on this.

Yes, I agree.

> I thought perhaps we could disable showing the popup menu when this feature was turned on, but that kind of seems like take a feature away without much gain. thoughts?

Not sure what you mean by this and have yet to test.

>     1. I have 2 points to make about the ctrl+click to return to "normal" playback speed. The first point is this, I originally had it to where a ctrl+click would return you back to 1x. Then I thought about it some more and decided that it would be better to return them to the default playback rate that they have selected in settings, which I think for most users will be the default 1x. What are yall's thoughts on this?

I personally believe reverting to the default playback rate is correct as well. A user's "normal" speed is their own choice!

>     2. Lastly, let's talk about how I am achieving the ctrl+click functionality.

My instinct is that all of this can be avoided with a simpler solution. For some pseudocode, an event listener to the effect of `video.on('click', (e) => {if (e.ctrlKey) { e.resetPlayback(); e.stopPropagation() }}) '` would work, no?

**Comment by codysechelski** [comment]
Thanks for the feedback..

> You will have to update the playbackRates array in ft-video-player.js with the new values.

When I do that, all possible playback rates that can be achieved using the scroll wheel will be listed in the popup menu that appears when you click the playback rate button. If I'm understanding your next comment. It seems like you agree that we should not show all of the possible rates. FYI, here's what it would look like
playback_rates

> You will also have to change the max value to 8 in player-settings.vue for default playback rate.
Great point. I didn't think about that. I will "make it happen"

In reference to my idea to not show the playback rate options while hovering over the playback button AND the "Scroll to change playback rate" setting is enabled:
> Not sure what you mean by this and have yet to test.

This is just a mock-up of my idea. I haven't researched enough to know how hard or easy it would be to implement. Maybe it's a bit wordy for a tooltip but as you may have guessed I tend to be overly verbose 😁
AEF187CC-05AD-4C7F-A4DC-DAB01EC949DF

> My instinct is that all of this can be avoided with a simpler solution. For some pseudocode, an event listener to the effect of video.on('click', (e) => {if (e.ctrlKey) { e.resetPlayback(); e.stopPropagation() }}) ' would work, no?
> 
> EDIT: If it's not "coming first", this post says you can do this with an inline onclick on the appropriate video element. The Vue idiomatic way to do this would be an @click="functionName($event)" on the target element, which I believe should work just as well.

So I tried that and did not have much luck.
ft-video-player.vue
```html
...
<video
      :id="id"
      class="ftVideoPlayer video-js vjs-default-skin dark"
      :poster="thumbnail"
      controls
      preload="auto"
      :data-setup="JSON.stringify(dataSetup)"
      @touchstart="handleTouchStart"
      @touchend="handleTouchEnd"
      @click="foo"
    >
...
```
ft-video-player.js
```javascript
    foo: function (event) {
      console.log('firing foo')
      console.log(event.ctrlKey)
      if (event.ctrlKey) {
        event.preventDefault()
        event.stopPropagation()
        console.log(this.player.paused())
      }
    },
```

The video still toggles play/pause with or without the ctrl key pressed. While the video is playing and I ctrl+click, `this.player.paused())` returns `true` which tells me that the player was paused before this event even fired.

But it does seem possible to fire a custom click handler before the play/pause action using the technique you mentioned. I made a codepen using vanilla JavaScript. So maybe I just need to keep fiddling with it, 
https://codepen.io/codysechelski/pen/JjJmYBX?editors=1111

However, even if I can get the order of execution to work in my favor, I STILL can't seem to stop the event from bubbling up to the play/pause toggle functionality, as you can see in the previously mention codepen.

If you are interested, here is a codepen that implements the "hack" I mentioned earlier in this PR
https://codepen.io/codysechelski/pen/vYZVLpa?editors=1111

I ended up just going to the discussion tab on video.js's GitHub page and asked the question there.
https://github.com/videojs/video.js/discussions/7447

Hopefully, someone there will answer, but I'm still "all ears" if anyone engaged with this project has any suggestions.

**Comment by jasonhenriquez** [comment]
> When I do that, all possible playback rates that can be achieved using the scroll wheel will be listed in the popup menu that appears when you click the playback rate button. If I'm understanding your next comment. It seems like you agree that we should not show all of the possible rates. FYI, here's what it would look like

Ah, I mean the same size intervals but up to 8.

> This is just a mock-up of my idea. I haven't researched enough to know how hard or easy it would be to implement. Maybe it's a bit wordy for a tooltip but as you may have guessed I tend to be overly verbose grin

I think it would be better to add logic to change the button to round to the nearest greater playbackRates interval value. e.g., 4.45 -> 4.5 on click).

**Comment by codysechelski** [comment]
Ok, so I changed the max play rate to 8 in the player settings. Also changed the pop-up menu on the player component to go up to 8 in steps of .25.

I don't want to come off as a DRY cultist, but is it worth moving the max playrate and the playrate "steps" to the settings file since I'm referencing them in 2 places?

Still waiting to hear back from my question on the video.js's GitHub page about the best way to add a click event that will not pause the video if the control key is pressed. so I guess we can just stand by on that one.

**Comment by jasonhenriquez** [comment]
> I don't want to come off as a DRY cultist, but is it worth moving the max playrate and the playrate "steps" to the settings file since I'm referencing them in 2 places?

I'm looking at your new function `playbackRates`, and I think this implementation is best kept as a hardcoded array of values for now. It reconstructs the array with the exact same values every time it is called. This programmatic generation of `playbackRates` is not useful as is, and although you could argue it might in the future, I fashion myself a YAGNI acolyte!

**Comment by codysechelski** [comment]
Roger that. 😀

**Comment by codysechelski** [comment]
Ok. Sorry @jasonhenriquez that I keep having to rope you in for help here, but here's where I'm at with the ctrl+click part...

I finally heard back from ma contributor to the video,js repo with some guidance (see )

I used that to mock it up in codepen. This is a working example of the proper way, according to the advice I received, to have a custom click event that has one set of logic when the ctrl key is pressed and still functions as expected when it's not.
https://codepen.io/codysechelski/pen/KKqGdGE?editors=0011

However, I immediately hit a roadblock when I tried to implement it on the ft-video-player component. The console is telling me that 

```javascript
this.player.tech(true)
```

is undefined.

Oh, I should mention that added that line to the `initializePlayer` function.

I will keep tinkering with it this weekend but I was just curious if you might know the reason this off the top of your head. THANKS!

**Comment by jasonhenriquez** [comment]
@codysechelski I'll take a look at it tonight & see what's up.

**Comment by codysechelski** [comment]
Thanks, @jasonhenriquez! 

I don't know if you had a chance to look at it or not, but  I kept digging through the issues on the video.js repo and found this: https://github.com/videojs/video.js/issues/2444. 

I was able to implement this solution in the codebase and test it out. It seems to work quite well. What do you think?

I left a comment in the code because I don't think it would be obvious as to why I am removing a click event only to add another one, but if you think it's unnecessary, I can remove it.

**Comment by codysechelski** [comment]
Just tested on my mac and realized I, at some point, removed `event.metaKey` from the `mouseScrollPlaybackRate` method. Bad copy/paste, I'm sure. Anyway, I added it back and tested everything on my Mac, and everything seems to be functioning as expected.

I was not able to test on Linux yet as I can't seem to build the app on my Ubuntu machine. I'm sure it's an issue with my environment. I just haven't had the time to tinker enough with it.

**Comment by codysechelski** [comment]
I have a question about the automatic checks...

Also, should the tests be rerun after a PR moved from "Draft" to "Ready"?

I pushed a commit. The tests went red because the "PR was not in the proper state". I then marked the PR as "Ready for Review", but the tests did not run again. So I ended up pushing a commit that did not make any "real" changes just to see if all of the tests will pass.

File: src/renderer/components/ft-video-player/ft-video-player.js
Comment (codysechelski) [inline src/renderer/components/ft-video-player/ft-video-player.js:481]: Just in case anyone is wondering about this, I had to modify the mouseSrollVolume method in order to make sure that the ctrl/command ket was NOT pressed so that a simple scroll will not change the volume AND the playback rate

File: src/renderer/components/ft-video-player/ft-video-player.js
Comment (codysechelski) [inline src/renderer/components/ft-video-player/ft-video-player.js:513]: Just in case anyone is wondering about this, I had to modify the mouseSrollVolume method in order to make sure that the ctrl/command ket was NOT pressed so that a simple scroll will not change the volume AND the playback rate

File: src/renderer/components/ft-video-player/ft-video-player.js
Comment (codysechelski) [inline src/renderer/components/ft-video-player/ft-video-player.js:921]: I was getting some rounding errors here

**Comment by jasonhenriquez** [comment]
> Im unable to test this. It keeps reloading the video in a loop without actually starting the video.
> 
> Can ur setting and Scroll volume over video player coexist? If not i would recommend that if ur setting is enabled and u want to enable the toggle scroll over volume, the other toggle needs to be disabled automatically. This goes for both directions

I disagree and think both settings can be active at the same time. However, a CTRL+Scroll should NOT trigger the audio scroll even when both are active. I have yet to test to see if this is how the PR handles this behavior currently. @codysechelski Thank you for your patience & we will eye this PR more closely after the coming release.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
I cant access the video for some reason. U can drag and drop ur video file into github just like screenshots. Only requirement is that it should be less then 10MB

**Comment by codysechelski** [comment]
> I cant access the video for some reason. U can drag and drop ur video file into github just like screenshots. Only requirement is that it should be less then 10MB

Cool. I guess I didn't know you could drop a video

https://user-images.githubusercontent.com/1571418/137607435-8931964f-6436-422d-82fd-0a64ea098142.mp4

**Comment by codysechelski** [comment]
Any more reviewers out there? If possible, I'd like to get this over the finish line so I can reallocate the sector in my brain that is holding the knowledge of what I did in the PR. 🙂

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
I requested reviewers for u but i dont know how bisy they are with life and stuff so please be a bit more patient.

On another note if u think this PR also fixes #616 and #1238 u can add them to the description

**Review by PikachuEXE — COMMENTED** [review]
My local test results: (with latest `development` merged into the branch locally)

> Adds the ability to use the ctrl key and mouse wheel to quickly change the playback rate.

It works but it changes playback rate by 0.05 per scroll
Is it intended? (Not supporting or objecting this idea)

> Changes the max playback rate from 3 to 8

Works
Though I am curious what can be used for 8...

> Adds the ability to use ctrl+click to return to the default playback rate.

Works

> Adds the ability to enable/disable the above in the settings > player settings screen (defaults to disabled)

Works and I like features introduced as opt-in

Will approve when my question on 1st point answered
Oh still need to review the code

**Comment by codysechelski** [comment]
Thanks for the review @PikachuEXE.

Yes, for point 1, that is the intended behavior. IMO, using your scroll wheel to change playback speed eventually turns it into a jog wheel and I believe the finer control is necessary.

On point 2: IMO this feature is intended to solve a different problem than being able to just set a playback speed with a point and a click. This existing functionality is good for just watching an entire video at 1.5x or 2x or something. However, using your scroll wheel to change the playback rate is more akin to a video editor using a jog wheel to scrub though a video quickly. It would not make sense to watch a whole video at 8x, but let's say that you were an expert at making pie crusts, but you wanted to know how to make an apple pie filling. If you found a video title "how to make an apple pie", then you could ctrl+scroll to quickly " fast forward" over the parts of the video that pertain to making the pie crust. Once you feel like you are getting close to the "good part", you can lower your speed for 8x to maybe 4 or 5x to sneak up up on it. Then when you see the apples being sliced you an ctrl+click to instantly return to 1x and learn all out that pie filling. 😀

So to me a max of 8x makes sense when you are using the scroll wheel, but it doesn't make as much sense in the playback speed popup menu on the tool bar. In fact, the first version of this PR left the that menu's max at 1.75, or 2, or whatever it was originally, I forget. But @jasonhenriquez thought it would be confusing to the users if the max playback rate was different in the two different contexts. And after I thought about it, I totally agree. So now the max is 8 in both places.

I'm hungry for pie. Now 🥧

**Review by PikachuEXE — APPROVED** [review]
Go make your pie 😃  🥧

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Havent tested this but just want to mention this

https://github.com/FreeTubeApp/FreeTube/issues/1273 when u hover over the playbackspeed icon does it start at 8?

@PikachuEXE tried to fix this behaviour in https://github.com/FreeTubeApp/FreeTube/pull/1274 but it didnt work out in the end.

If the same behaviour also occurs i think we need to solve it first because nobody would like to scroll down from 8 to 2 for example while u hover on this list

**Comment by PikachuEXE** [comment]
[oof](https://www.merriam-webster.com/dictionary/oof)
image

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
*sad noises*

**Comment by absidue** [comment]
> 
> 
> Havent tested this but just want to mention this
> 
> #1273 when u hover over the playbackspeed icon does it start at 8?
> 
> @PikachuEXE tried to fix this behaviour in #1274 but it didnt work out in the end.
> 
> If the same behaviour also occurs i think we need to solve it in some kind of way first because nobody would like to scroll down from 8 to 2 for example while u hover on the list
> 
> Edit: pls accept this wonderful piece of cake 🍰

I've just opened #1902 which focuses the current playback speed when the menu is opened, which fixes this issue.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
U are a hero!
I will test asap.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Im unable to test till sunday so
One more question before the last approval is given.

In the settings page when u want to alter the default playback speed is it also possible to move the slider to 8x in steps of 0.1?

**Comment by codysechelski** [comment]
Hey @efb4f5ff-1298-471a-8973-3d47447115dc , sorry I just was your latest comment,
The `Default Playback Rate` setting in the `Player Settings` section can be set in increments of `.25`.

This was the case before this PR and I did not change it. It's also the same increments that the playback rate button uses in the video player's toolbar. IMO, that is the correct increment for this setting.

While it may seem strange that there are playback rates that are reachable using ctrl+scroll that are not reachable in the playback rate button or the default playback settings slider, I think the greater granularity (.05) in the ctrl+scroll feature is merited while "jogging" through the video but would be overwhelming for users in the other contexts.

**Comment by PikachuEXE** [comment]
Since "scroll to adjust playback speed" a new feature I don't object on having 0.05 per scroll
Can adjust that later 

I also agree that there is no need to have every value with 0.05 interval in default playback speed setting

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Thanks for ur explanation. At first it indeed looked a little strange to me but after ur rundown I totally agree with u!

Hopefully a last reviewer will review this soon.

In the meantime have some 🍰

**Review by ChunkyProgrammer — APPROVED** [review]
LGTM

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Hi @codysechelski, we are currently fixing https://github.com/FreeTubeApp/FreeTube/issues/6668 in PR https://github.com/FreeTubeApp/FreeTube/pull/6943 but i noticed something weird and wanted to ask you if this is intended behavior.

Default playback rate is set to 1x in the setting -> playback rate has been changed while watching the video, 9x -> next video uses playback rate of previous video so still 9x-> ctrl click resets to 9x instead of the default playback rate predefined in the settings


https://github.com/user-attachments/assets/62759840-7677-4335-9ed4-8b38251f6c99

**Pull request merged**
